keps/sig-cli/20190318-kustomize-generators-transformers.md

```diff
@@ -0,0 +1,435 @@
+   - Group matches the plugin name
+   - Exits non-0 if no plugins are found any generator entries
+1. For each Virtual Resource *Group*
+  1. Kustomize execs the plugin `transform` command
+  1. Kustomize writes the Virtual Resources in that Group to the process STDIN
+  1. Kustomize writes **all** non-Virtual Resources to the process STDIN
+  1. Kustomize reads the set of transformed Resources from the process STDOUT
+  1. Kustomize reads error messages from the exec process STDERR
+  1. Kustomize fails if the plugin exits non-0
+  1. Kustomize replaces its current set of non-Virtual Resources with the set of emitted Resources.
+
+The order of plugin execution is arbitrary.
+
+##### Restrictions
+
+For the initial iteration, transformers cannot add/remove Resources, or change their names/namespaces.
+The ability doing so would be significantly make complex ordering interactions much more likely.  E.g. transformers
+would need to keep and propagate transformations from other transformers.
+
+We are prioritizing a restrictive but predictable API over a powerful but complex one.
+
+### Phases
+
+Follow is the Kustomize workflow:
+
+1. Read all `generators` and `transformers`
+1. Read all `patches`
+1. Apply Virtual Resource `patches` to `generators` and `transformers`
+1. Generate non-Virtual Resource set from Built-In Generators `inputs`, `bases`, `secretGenerator`, etc
+1. Generate non-Virtual Resources from Plugin-Generators and add to non-Virtual Resource set
+1. Transform non-Virtual Resources using Plugin-Generators
+1. Transform non-Virtual Resources using Built-In Generators (including a second round of `patches`)
+1. Built-In Sorting
+
+### User Stories [optional]
+
+#### DeclarativeEcosystem Solution Plugins
+
+Alice uses GitOps for deployment of her Kubernetes Resource Config.  Alice has a Helm
+chart that she would like to include into her GitOps workflow.  Alice would like to be
+able to check the chart and values.yaml into her repository and have it deployed
+just like her other config.  Alice would also like to be able to apply cross cutting
+transformations - such as labels, annotations, name-prefixes, etc - to the Resources generated
+by the Helm chart.
+
+Alice uses the Kustomize helm chart plugin to declaratively generate
+Resource Config using Helm charts, which will then have kustomizations applied.
+
+1. Alice downloads and installs the Helm generator plugin from *github.com/kubernetes-sigs/kustomize/*
+   and installs it at `$XDG_CONFIG_HOME/kustomize/plugins/generators/helm.kustomize.io`.
+1. Alice creates the `chart.yaml` Resource Config and adds it to her `kustomization.yaml` as a `generators` entry.
+  - Having the groupVersion `helm.kustomize.io` triggers the generator under `plugins/generators/helm.kustomize.io`
+
+```yaml
+groupVersion: helm.kustomize.io/v1beta1
+kind: Chart
+generate:
+    chart: ./path/to/chart/from/kustomization.yaml
+    values:
+      k1: v1
+      k2: v2
+```
```

Comment:
plugin can read/write arbitrary files for ex. values.yaml ? (otherwise it will need some sandboxing mechanism).

Reply:
Plugins are executables, and can be:

- Malicious
- Exploitable

We might be able to provide some sandboxing by running them in a container...